### PR TITLE
Added the purge API on node endpoints

### LINKED ISSUE #3444: API for purging nodes
We need an API to purge nodes on the servers. The use case is we run Nomad on bare metal machines and sometimes when the machine fails because of hardware issues and is fixed by data-center operators we have to re-image the machine and have them join back in. The machines have the same host name and IP addresses so re-registration with servers fail because of mismatch in secret IDs.

I am thinking an API to purging the node in the server should help in this case. The security side of things are handled in our case by secure introduction of the machines(can't go into the details but a machine can't simply appear in the DC or join a cluster).

@dadgar want to weigh in your thoughts? I can work on the implementation since we need this right away.

### PR DESCRIPTION
**Pull request by diptanu**
Fixes #3444 

Tested by creating a client and calling the purge API. The server removed the client, however, the client could register if it was still alive which I believe is the expected behavior.

I will add the http api docs on the website once the PR is merged.

### CODE REVIEW

File: nomad/node_endpoint.go [hunk 1]
```diff
@@ -232,7 +246,6 @@ func (n *Node) Deregister(args *structs.NodeDeregisterRequest, reply *structs.No
 	}
 
 	// Determine if there are any Vault accessors on the node
-	ws := memdb.NewWatchSet()
 	accessors, err := n.srv.State().VaultAccessorsByNode(ws, args.NodeID)
```
Comment (diptanu) [inline nomad/node_endpoint.go]: @dadgar Can we use the snap we created earlier?

Comment (dadgar) [reply nomad/node_endpoint.go]: Yeah you can switch it over

File: nomad/node_endpoint.go [hunk 2]
```diff
@@ -213,6 +213,20 @@ func (n *Node) Deregister(args *structs.NodeDeregisterRequest, reply *structs.No
 	if args.NodeID == "" {
 		return fmt.Errorf("missing node ID for client deregistration")
 	}
+	// Look for the node
```
Comment (dadgar) [inline nomad/node_endpoint.go]: Before the `// Verify the arguments` section can you add:

```
	// Check node write permissions
	if aclObj, err := n.srv.ResolveToken(args.AuthToken); err != nil {
		return err
	} else if aclObj != nil && !aclObj.AllowNodeWrite() {
		return structs.ErrPermissionDenied
	}
```

Comment (dadgar) [reply nomad/node_endpoint.go]: We will need to add a test similar to this: https://github.com/hashicorp/nomad/blob/master/nomad/node_endpoint_test.go#L613

What you need to test is that an ACL with node write/management token work and that another token does not.

File: command/agent/node_endpoint.go [hunk 3]
```diff
@@ -42,6 +42,9 @@ func (s *HTTPServer) NodeSpecificRequest(resp http.ResponseWriter, req *http.Req
 	case strings.HasSuffix(path, "/drain"):
 		nodeName := strings.TrimSuffix(path, "/drain")
 		return s.nodeToggleDrain(resp, req, nodeName)
+	case strings.HasSuffix(path, "/purge"):
```
Comment (dadgar) [inline command/agent/node_endpoint.go]: HTTP Documentation

Comment (diptanu) [reply command/agent/node_endpoint.go]: Done

**Comment by diptanu** [comment]
Addressed all the comments.

File: command/agent/node_endpoint.go [hunk 4]
```diff
@@ -142,3 +145,19 @@ func (s *HTTPServer) nodeQuery(resp http.ResponseWriter, req *http.Request,
 	}
 	return out.Node, nil
 }
+
+func (s *HTTPServer) nodePurge(resp http.ResponseWriter, req *http.Request, nodeID string) (interface{}, error) {
+	if req.Method != "POST" {
```
Comment (dadgar) [inline command/agent/node_endpoint.go]: I am going to merge and also allow PUT

**Pull request merged**